### PR TITLE
Finish pr 13

### PR DESCRIPTION
**Pull request by dunglas**
Finish #13  (closed by mistake). /cc @vincentchalamon

### CODE REVIEW

File: src/Resources/views/Collector/mercure.html.twig [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        <div class="empty">
+            <p>No messages have been collected.</p>
+        </div>
+    {% else %}
+        <div class="sf-tabs">
+            {% for hub, publisher in collector.publishers %}
+                <div class="tab">
+                    <h3 class="tab-title">{{ hub }}<span class="badge">{{ publisher.count }}</span></h3>
+                    <div class="tab-content">
+                        <p class="text-muted">{{ hub }}</p>
```
Comment (vincentchalamon) [inline src/Resources/views/Collector/mercure.html.twig]: It's useless to display the hub name here, as it's already displayed on the tab name (line 36)

File: src/DataCollector/MercureDataCollector.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        return 'mercure';
+    }
+
+    public function count(): int
+    {
+        return $this->data['count'];
+    }
+
+    public function getDuration(): int
```
Comment (chalasr) [inline src/DataCollector/MercureDataCollector.php]: should be `: float`

File: src/DataCollector/MercureDataCollector.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            'count' => 0,
+            'duration' => 0,
+            'memory' => 0,
+            'publishers' => $this->publishers,
+        ];
+
+        foreach ($this->publishers as $name => $publisher) {
+            $this->data['duration'] += $publisher->getDuration();
+            $this->data['memory'] += $publisher->getMemory();
+            $this->data['count'] += count($publisher->getMessages());
```
Comment (chalasr) [inline src/DataCollector/MercureDataCollector.php]: `\count`

File: src/DependencyInjection/MercureExtension.php [hunk 4]
```diff
@@ -67,9 +74,31 @@ public function load(array $configs, ContainerBuilder $container)
...
+                $publishers[$name] = new Reference("$hubId.traceable");
+                if ($defaultHub === $hubId) {
+                    $container->setAlias(TraceablePublisher::class, "$hubId.traceable");
+                }
+            }
+        }
+
+        if ($enableProfiler) {
+            $container->register('data_collector.mercure', MercureDataCollector::class)
+                ->addArgument($publishers)
```
Comment (chalasr) [inline src/DependencyInjection/MercureExtension.php]: I suggest `new IteratorArgument($publishers);`

File: composer.json [hunk 5]
```diff
@@ -17,10 +17,10 @@
     ],
     "require": {
         "php": "^7.1.3",
-        "symfony/config": "^3.4|^4.0|^5.0",
-        "symfony/dependency-injection": "^3.4|^4.0|^5.0",
-        "symfony/http-kernel": "^3.4|^4.0|^5.0",
-        "symfony/mercure": "^0.2"
+        "symfony/config": "^4.3|^5.0",
```
Comment (vincentchalamon) [inline composer.json]: `symfony/*` v3.4 is still active, why removing its compatibility?

Comment (dunglas) [reply composer.json]: Because we use 4.0 only features.

Comment (vincentchalamon) [reply composer.json]: So this version has a big BC break :disappointed: !

What are the 4.0 only features used in this PR?

Comment (vincentchalamon) [reply composer.json]: Isn't it possible to handle the `setDeprecated` method differently, as it's the only 4.0 feature added in this PR?

Comment (dunglas) [reply composer.json]: Bumping deps isn't a BC break, but I'm not against supporting 3.4.

File: composer.json [hunk 6]
```diff
@@ -17,10 +17,11 @@
     ],
     "require": {
         "php": "^7.1.3",
-        "symfony/config": "^3.4|^4.0|^5.0",
-        "symfony/dependency-injection": "^3.4|^4.0|^5.0",
-        "symfony/http-kernel": "^3.4|^4.0|^5.0",
-        "symfony/mercure": "^0.2"
+        "symfony/config": "^3.4|^4.3|^5.0",
```
Comment (vincentchalamon) [inline composer.json]:
```suggestion
        "symfony/config": "^3.4|^4.0|^5.0",
```

File: composer.json [hunk 7]
```diff
@@ -17,10 +17,11 @@
...
+        "symfony/contracts": "^1.1.7|^2",
+        "symfony/dependency-injection": "^3.4|^4.3|^5.0",
```
Comment (vincentchalamon) [inline composer.json]:
```suggestion
        "symfony/dependency-injection": "^3.4|^4.0|^5.0",
```

File: composer.json [hunk 8]
```diff
@@ -17,10 +17,11 @@
...
+        "symfony/http-kernel": "^3.4|^4.3|^5.0",
```
Comment (vincentchalamon) [inline composer.json]:
```suggestion
        "symfony/http-kernel": "^3.4|^4.0|^5.0",
```

File: src/DependencyInjection/MercureExtension.php [hunk 9]
```diff
@@ -67,9 +74,35 @@ public function load(array $configs, ContainerBuilder $container)
...
+
+            if ($enableProfiler) {
+                $container->register("$hubId.traceable", TraceablePublisher::class)
+                    ->setDecoratedService($hubId)
+                    ->addArgument(new Reference("$hubId.traceable.inner"))
+                    ->addArgument(new Reference('debug.stopwatch'));
+
+                $publishers[$name] = new Reference("$hubId.traceable");
+                if ($defaultHub === $hubId) {
+                    $container->setAlias(TraceablePublisher::class, "$hubId.traceable");
```
Comment (chalasr) [inline src/DependencyInjection/MercureExtension.php]: do we need this alias? I mean, nobody should typehint against `TraceablePublisher` and the `PublisherInterface` type should naturally be wired with the traceable default hub thanks to decoration, or am I mistaken?

File: composer.json (hunk 6 above)
Comment (chalasr) [reply composer.json]: 4.x releases prior to 4.3 are not maintained anymore, not sure we should care

**Comment by dunglas** [comment]
Thanks @vincentchalamon! And thanks @chalasr for the review.

**Pull request merged**